pkg/spoke/submarineragent/diagnose_controller.go

```diff
@@ -0,0 +1,219 @@
+	v1 "k8s.io/api/core/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+
+	corev1lister "k8s.io/client-go/listers/core/v1"
+	addoninformerv1alpha1 "open-cluster-management.io/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "open-cluster-management.io/api/client/addon/listers/addon/v1alpha1"
+)
+
+const (
+	jobName = "submariner-diagnose"
+	imageName = "quay.io/vthapar/submariner-diagnose:latest"
+	serviceAccount = "submariner-operator"
+	jobLabel = "job-name"
+)
+
+// submarinerDiagnoseConfigController watches the SubmarinerDiagnoseConfigs API on the hub cluster
+// and run Diagnose Job on the manged cluster.
+type submarinerDiagnoseConfigController struct {
+	kubeClient           kubernetes.Interface
+	diagnoseClient       diagnoseclient.Interface
+	addOnLister          addonlisterv1alpha1.ManagedClusterAddOnLister
+	diagnoseConfigLister diagnoselister.SubmarinerDiagnoseConfigLister
+	podLister            corev1lister.PodLister
+	clusterName          string
+	onSyncDefer          func()
+}
+
+type SubmarinerDiagnoseConfigControllerInput struct {
+	ClusterName            string
+	KubeClient             kubernetes.Interface
+	ConfigClient           diagnoseclient.Interface
+	AddOnInformer  addoninformerv1alpha1.ManagedClusterAddOnInformer
+	ConfigInformer diagnoseinformer.SubmarinerDiagnoseConfigInformer
+	PodInformer    corev1informers.PodInformer
+	Recorder       events.Recorder
+	// This is a hook for unit tests to invoke a defer (specifically GinkgoRecover) when the sync function is called.
+	OnSyncDefer func()
+}
+
+// NewSubmarinerDiagnoseConfigController returns an instance of submarinerAgentConfigController.
+func NewSubmarinerDiagnoseConfigController(input *SubmarinerDiagnoseConfigControllerInput) factory.Controller {
+	c := &submarinerDiagnoseConfigController{
+		kubeClient:           input.KubeClient,
+		diagnoseClient:       input.ConfigClient,
+		addOnLister:          input.AddOnInformer.Lister(),
+		diagnoseConfigLister: input.ConfigInformer.Lister(),
+		podLister: input.PodInformer.Lister(),
+		clusterName:          input.ClusterName,
+		onSyncDefer:          input.OnSyncDefer,
+	}
+
+	return factory.New().
+		WithFilteredEventsInformers(func(obj interface{}) bool {
+			metaObj := obj.(metav1.Object)
+
+			return metaObj.GetName() == constants.SubmarinerAddOnName
+		}, input.AddOnInformer.Informer()).
+		WithFilteredEventsInformers(func(obj interface{}) bool {
+			metaObj := obj.(metav1.Object)
+
+			return metaObj.GetName() == constants.SubmarinerConfigName
+		}, input.ConfigInformer.Informer()).
+		WithFilteredEventsInformers(func(obj interface{}) bool {
+			metaObj := obj.(metav1.Object)
+			if jobLabelValue, has := metaObj.GetLabels()[jobLabel]; has {
+				if jobLabelValue == jobName {
+					return true
+				}
+			}
```

Comment:
It seems expensive to watch all pods. Perhaps we can watch Jobs instead?

Reply:
Yeah, that was my concern too, how does this filtered informer work? If it is implemented in backend as a single informer for all controllers and just notifies us all of events, that should be fine. Jobs won't work as we need the podname and can't be sure if pod is available by the time we get Job notification. Is there a way to watch for pods within a specific namespace only? Alternately, would it be okay to poll for pod creation?

Reply:
I think you can watch for the Job and query for the associated pod as you're doing. If it doesn't exist then requeue. Or just wait for the next Job notification. I assume you want to update the status condition when the Job completes? If so then perhaps check the Job status rather the querying the pod.

Reply:
Getting back to it with 2.5 delayed.

@tpantelis We don't even need to wait for Job completion, just need the name of the pod so in UI we can link to that pod's logs. UI can stream logs in realtime, which works even better as user can see the progress of diagnose from the logs.

Reply:
Revisiting this.... I'm still wary of the overhead of watching all pods. I still think it would be better to watch for the `Job` and query for the associated pod as outlined above.